examples/star-wars/js/components/StarWarsApp.js

```diff
@@ -13,23 +13,72 @@
 import React from 'react';
 import Relay from 'react-relay';
 import StarWarsShip from './StarWarsShip';
+import AddShipMutation from '../mutation/AddShipMutation';
 
 class StarWarsApp extends React.Component {
+  constructor() {
+    super();
+    this.state = {
+      shipName: '',
+      factionId: 1
+    };
+  }
+
+  addship() {
+    var name = this.state.shipName;
+    Relay.Store.update(
+      new AddShipMutation({name, faction: this.props.factions[this.state.factionId]})
+    );
+  }
+
+  handleInputChange(e) {
+    this.setState({
+      shipName: e.target.value
+    });
+  }
+
+  handleSelectionChange(e) {
+    this.setState({
+      factionId: e.target.value
+    });
+  }
+
   render() {
     var {factions} = this.props;
     return (
-      <ol>
-        {factions.map(faction => (
-          <li>
-            <h1>{faction.name}</h1>
-            <ol>
-              {faction.ships.edges.map(edge => (
-                <li><StarWarsShip ship={edge.node} /></li>
-              ))}
-            </ol>
-          </li>
-        ))}
-      </ol>
+      <div>
+        <ol>
+          {factions.map(faction => (
+            <li>
+              <h1>{faction.name}</h1>
+              <ol>
+                {faction.ships.edges.map(edge => (
+                  <li><StarWarsShip ship={edge.node} /></li>
+                ))}
+              </ol>
+            </li>
+          ))}
+            <li>
+              <h1>Introduce Ship</h1>
+              <ol>
```

Comment:
Rather than change tracking the text input, what do you think about a simpler design where this is made a bonafide `<form>` with an `onSubmit` handler. In your `handleAddShipSubmit` method, you can simply pull the ship name and faction from refs, and then reset the form.

```js
handleAddShipSubmit(e) {
  e.preventDefault();
  Relay.Store.update(
    new AddShipMutation({
      faction: this.props.factions[this.refs.factionIdSelect.value],
      name: this.refs.shipNameInput.value,
    })
  );
  this.refs.shipNameInput.value = '';
}
```

```js
<form onSubmit={this.handleAddShipSubmit.bind(this)}>
  <input type="text" ref="shipNameInput" />
  <select ref="shipNameInput" />…</select>
</form>
```